server/etcdserver/server.go

```diff
@@ -1050,6 +1048,10 @@ func (s *EtcdServer) applySnapshot(ep *etcdProgress, apply *apply) {
 	ep.appliedi = apply.snapshot.Metadata.Index
 	ep.snapi = ep.appliedi
 	ep.confState = apply.snapshot.Metadata.ConfState
+
+	// As backends and implementations like alarmsStore changed, we need
+	// to re-bootstrap Appliers.
+	s.uberApply = newUberApplier(s)
```

Comment:
Instead of storing a alarmStore reference, could we store reference to alarmStore getter interface?

Reply:
I would agree in general, but it would require deep reform what `Backend` is and how restore from snapshot works. 

Currently restore from snapshot: 
  - creates a new instance of the backend
  - all the objects/helpers pointing on the all backend are 'poinsoned'. 
  - we need to rebootstrap them using the new backend

It might be easier / safer to create them from scratch rather make sure there is `foo.RebootstrapFromBackend(BE)` method that will entirely clean it's state and will load what needed from `be`

Reply:
I think we can revisit this comment in a separate PR if needed.